Feature/volumeaxis offset

martinfalk opened the pull request:
Changes proposed in this PR:
 * relative offsets for labels and captions in `VolumeAxis`
 * modules sorted by name in Tools menu
 * added help for axis related properties
 * typo/warn fix

Relative scaling of axis labels and captions as well as tick lengths ensures reasonable axes when the volume basis changes. In the newly added regression test (see image), the basis of the volume is scaled by 1, 10000, and 0,00001 while using the same axis offsets.

![image](https://user-images.githubusercontent.com/9251300/207404656-df4a5d2e-1df2-45e3-ac87-ef48bf4e9810.png)


This has been tested on: Windows/MSVC 2022

martinfalk committed:
FontRendering,Plotting,PlottinGL: help update

martinfalk committed:
Core,Base,QtWidgets,Python3,Tools: typo fix

martinfalk committed:
Core: warn fix

martinfalk committed:
Plotting,PlottingGL: added scaling factor to AxisSettings

* the scaling factor affects caption and label offsets as well as tick lengths for relative scaling

martinfalk committed:
PlottingGL: VolumeAxis relative offset scaling

martinfalk committed:
Testing: regression test of VolumeAxis for scaled volumes

martinfalk committed:
Qt: sort modules in Tools menu

ResearchDaniel commented on modules/plotting/src/properties/axisproperty.cpp:
```diff
@@ -67,34 +68,53 @@ class PlotTextSettings;
-    , flipped_{"flipped", "Swap Label Position", false}
+    , range_{"range",
+             "Axis Range",
+             "Current range of the axis"_help,
```
```suggestion
             "Value range of the axis"_help,
```

ResearchDaniel commented on modules/plotting/src/properties/categoricalaxisproperty.cpp:
```diff
@@ -64,32 +64,34 @@ CategoricalAxisProperty::CategoricalAxisProperty(
-    addProperty(flipped_);
-    addProperty(orientation_);
-    addProperty(placement_);
+    scalingFactor_.setVisible(1.0f);
```
```suggestion
    scalingFactor_.setVisible(false);
```

ResearchDaniel commented on modules/plottinggl/src/processors/volumeaxis.cpp:
```diff
@@ -161,18 +162,29 @@ const ProcessorInfo VolumeAxis::processorInfo_{
     CodeState::Stable,        // Code state
     "GL, Plotting",           // Tags
-};
+    "Renders an x, y, and z axis next to the input volume."_help};
```
```suggestion
    "Renders x, y, and z axes next to the input volume."_help};
```

ResearchDaniel commented on modules/plottinggl/src/processors/volumeaxis.cpp:
```diff
@@ -385,6 +402,36 @@ void VolumeAxis::process() {
+                case OffsetScaling::MeanExtent:
+                    return glm::compAdd(l) / (3.0f * 100.0f);
+                case OffsetScaling::Diagonal:
+                    return glm::compAdd(l) / 100.0f;
```
Should be sqrt(3)*100?

martinfalk replied:
Should read `glm::length(m[0]+m[1]+m[2])` instead.

ResearchDaniel commented on modules/plotting/include/modules/plotting/datastructures/axissettings.h:
```diff
@@ -57,6 +57,8 @@ class IVW_MODULE_PLOTTING_API AxisSettings {
     virtual vec4 getColor() const = 0;
     virtual float getWidth() const = 0;
+    // scaling factor affecting tick lengths and offsets of axis caption and labels
+    virtual float getScalingFactor() const = 0;
```
Factor is redundant. Could name it scaling, transform, or offsetScaling. OffsetScaling seem to be used in later.
```suggestion
    virtual float getOffsetScaling() const = 0;
```

petersteneteg replied:
Since it affects more than the offset, according to the comment, I think getScalingFactor or getScaleFactor is appropriate

ResearchDaniel commented on modules/plottinggl/src/processors/volumeaxis.cpp:
```diff
@@ -161,18 +162,29 @@ const ProcessorInfo VolumeAxis::processorInfo_{
+    , imageInport_{"imageInport", "Background image (optional)"_help}
+    , outport_{"outport",
+               "Output image containing the rendered volume axes and the optional input image"_help}
+    , offsetScaling_{"offsetScaling",
```
What would be the reason for choosing one over the other here? If all options are necessary we might want to provide some guidelines in the help.

martinfalk replied:
Assuming you have a rather uniform basis, then it doesn't matter. However when loading volumes with rather different dimensions these options have different effects. Think of narrow vs. large volumes in a time series.

martinfalk replied:
If you load volumes which grow in one dimension, you probably want min to be consistent.
 If the smallest side is shrinking you don't want that one.

martinfalk replied:
Added the following help
```c++
                     R"(Offset scaling affects tick lengths and offsets of axis captions and labels.
                        + *None* No scaling, offsets and lengths are given in world coordinates.
                        + *Min* Relative scaling based on the shortest extent of the volume. Useful 
                                when visualizing a growing volume.
                        + *Max* Relative scaling based on the longest extent of the volume. Useful 
                                when visualizing a shrinking volume.
                        + *Mean* Relative scaling basd on the mean volume extent.
                        + *Diagonal* Relative scaling based on the diagonal of the volume.
                      )"_unindentHelp,
```
Might consider to either use `mean` or `diagonal`, if we want to reduce the number of options.

ResearchDaniel replied:
Would it be possible to choose this automatically based on these guidelines with manual override?

petersteneteg replied:
I don't think it is worth the effort to try and automate that, just use mean or diagonal as default.

ResearchDaniel reviewed (dismissed):
Looks good overall, but have some questions and suggestions. The axes labels in the regression test seem to overlap the tick text. Is it possible avoid this? If not, is it possible to change the workspace so that it does not?

martinfalk committed:
Plotting,PlottingGL: requested changes

martinfalk committed:
Plotting,PlottingGL: default values

martinfalk committed:
Testing: updated axis regression tests

martinfalk committed:
PlottingGL: label default alignment

martinfalk committed:
Testing: updated plottinggl and dataframe regression tests

martinfalk committed:
Meta: Changelog update

petersteneteg commented on modules/plotting/src/properties/axisproperty.cpp:
```diff
@@ -52,10 +52,12 @@
 #include <fmt/core.h>    // for basic_string_view
 #include <fmt/printf.h>  // for sprintf
 #include <glm/vec2.hpp>  // for vec<>::(anonymous)
+#include <glm/gtx/vec_swizzle.hpp>
```
unused?

martinfalk replied:
Used in lines 283/284
```c++
        labelSettings_.font_.anchorPos_.set(glm::yx(labelSettings_.font_.anchorPos_.get()));
        captionSettings_.font_.anchorPos_.set(glm::yx(captionSettings_.font_.anchorPos_.get()));
```

petersteneteg commented:
modules/plottinggl/tests/regression/axiscombinationtest/VolumeAxis2.png
seems to have gone for the worse I think, The units are lost? and the labels are 2 close to the axes, I the new one is the new defaults, they should be further away, especially in the 2d plot

martinfalk committed:
Plotting,PlottingGL: revised AxisSettings

* moved placement of labels (inside/outside) from axis to axis labels and caption

martinfalk committed:
PlottingGL: buttons for axis orientation and alignment

martinfalk committed:
PlottingGL: hide axis orientation in VolumeAxis

martinfalk committed:
PlottingGL: revised axis label rotation

martinfalk committed:
PlottingGL: refactored ColorScaleLegend processor

* added buttons with orientation/position presets

martinfalk committed:
PlottingGL: option typo in parallel coordinates plot

martinfalk committed:
PlottingGL: PCP margin fix

martinfalk committed:
PlottingGL: blending set to pre-multiplied in PCP

* was inconsistent with all other plotting and text rendering
* optional background inport

martinfalk committed:
PlottingGL: PCP antialiasing for lines

martinfalk committed:
PlottingGL: scatter plot update fix

martinfalk committed:
PlottingGL: updated example workspaces

martinfalk committed:
Testing: updated regression tests PlottingGL, DataFrame

martinfalk committed:
Core: CompositeProperty serialization fix

martinfalk committed:
BaseGL: Mesh2DRenderProcessorGL drawing in screen coords

* use the dimensions of the outport to define the view frustum

martinfalk committed:
Testing: updated regression tests PlottingGL, DataFrame

martinfalk committed:
PlottingGL: format/warn fix

martinfalk commented:
Revised the internal state of the `AxisProperty` and `ColorScaleLegend`. Presets for axis placement and label alignment are available as buttons.

Additional changes:
* switched blending in PCP to pre-multiplied to be consistent with all other plots, text rendering, and the default blending in the background processor
* label and caption rotation in the AxisProperty are now handled identically
* added functionality to `2D Mesh Renderer` processor to use the Image of the outport for determining the view frustum. This allows rendering geometry directly in screen coordinates.

![image](https://user-images.githubusercontent.com/9251300/217891395-932f6269-34cd-4b87-8656-981980a96182.png)

**2D Mesh Renderer drawing axis bounding boxes using screen coordinates**
<img src='https://user-images.githubusercontent.com/9251300/217894736-3e4eb8b1-3c56-404b-bfd6-0ee18aee4158.png' width='300px'><img src='https://user-images.githubusercontent.com/9251300/217892087-84fb740e-665d-422a-a30e-fccd300acca2.png' width='400px'>

petersteneteg commented on modules/plotting/include/modules/plotting/datastructures/majorticksettings.h:
```diff
@@ -50,8 +50,17 @@ class IVW_MODULE_PLOTTING_API MajorTickSettings {
 IVW_MODULE_PLOTTING_API bool operator==(const MajorTickSettings& a, const MajorTickSettings& b);
 IVW_MODULE_PLOTTING_API bool operator!=(const MajorTickSettings& a, const MajorTickSettings& b);
+
+namespace utilplot {
```
Do we use utilplot anywhere else, otherwise I would just use util, or if you really want a different namespace I would just go with plot

petersteneteg commented on modules/plotting/include/modules/plotting/datastructures/plottextsettings.h:
```diff
@@ -56,6 +59,13 @@ class IVW_MODULE_PLOTTING_API PlotTextSettings {
+/**
+ * flip inside and outside direction of placement \p p
+ */
+IVW_MODULE_PLOTTING_API LabelPlacement flip(LabelPlacement p);
```
I think Inviwo::plot::untilplot::flip is a bit overkill

petersteneteg commented on modules/plotting/include/modules/plotting/properties/tickproperty.h:
```diff
@@ -57,6 +57,9 @@ class IVW_MODULE_PLOTTING_API MajorTickProperty : public MajorTickSettings,
     virtual std::string getClassIdentifier() const override;
     static const std::string classIdentifier;
 
+    MajorTickProperty(std::string_view identifier, std::string_view displayName, Document help,
```
Im not sure If we need to pass in individual help text into each Composite property here. Maybe it is better just to hardcode the help text in the Constructor of the property...

petersteneteg commented on modules/plotting/src/properties/axisproperty.cpp:
```diff
@@ -112,56 +132,88 @@ AxisProperty::AxisProperty(std::string_view identifier, std::string_view display
+AxisProperty::AxisProperty(std::string_view identifier, std::string_view displayName,
+                           Orientation orientation, bool includeOrientationProperty,
+                           InvalidationLevel invalidationLevel, PropertySemantics semantics)
+    : AxisProperty{identifier,        displayName, {}, orientation, includeOrientationProperty,
```
If you are updating the help added a default help text here would be good...

petersteneteg commented on modules/plotting/src/properties/axisproperty.cpp:
```diff
@@ -112,56 +132,88 @@ AxisProperty::AxisProperty(std::string_view identifier, std::string_view display
-    range_.setReadOnly(useDataRange_.get());
-    useDataRange_.onChange([this]() { range_.setReadOnly(useDataRange_.get()); });
+    // insert orientation property only if rhs owns one, too
+    if (rhs.getPropertyByIdentifier(rhs.orientation_.getIdentifier())) {
```
might wrap the orientation one in an std::optional?

petersteneteg commented on modules/plotting/src/properties/axisstyleproperty.cpp:
```diff
@@ -55,18 +55,27 @@ AxisStyleProperty::AxisStyleProperty(std::string_view identifier, std::string_vi
                                      InvalidationLevel invalidationLevel,
                                      PropertySemantics semantics)
     : CompositeProperty(identifier, displayName, invalidationLevel, semantics)
```
Might add a help text here to...

petersteneteg commented on modules/plotting/src/properties/plottextproperty.cpp:
```diff
@@ -46,47 +46,59 @@ const std::string PlotTextProperty::classIdentifier = "org.inviwo.PlotTextProper
+PlotTextProperty::PlotTextProperty(std::string_view identifier, std::string_view displayName,
+                                   bool checked, InvalidationLevel invalidationLevel,
+                                   PropertySemantics semantics)
+    : PlotTextProperty(identifier, displayName, {}, checked, invalidationLevel, semantics) {}
```
might add help text here

petersteneteg commented on modules/plotting/src/properties/tickproperty.cpp:
```diff
@@ -167,17 +170,17 @@ TickProperty::TickProperty(std::string_view identifier, std::string_view display
-    addProperty(majorTicks_);
-    addProperty(minorTicks_);
+
+    addProperties(majorTicks_, minorTicks_);
```
Could be just get rid of this property? Would be nice to not have to unfold that level since it is only 2 nodes in it?

martinfalk replied:
Removed this property since it was used nowhere.

petersteneteg reviewed (dismissed):
Nice Work! looking forward to trying it out!

martinfalk committed:
Plotting: requested changes

martinfalk committed:
Meta: Changelog update

martinfalk committed:
PlottingGL: remove label overlap in regression test

martinfalk commented:
@ResearchDaniel 
> Looks good overall, but have some questions and suggestions. The axes labels in the regression test seem to overlap the tick text. Is it possible avoid this? If not, is it possible to change the workspace so that it does not?

Resolved all questions and addressed the overlap in the regression test.

![image](https://user-images.githubusercontent.com/9251300/218414235-04ad1144-737e-4f2e-9bb1-a35e5bd6a31d.png)
